File: src/services/storageService.js
import { ref, uploadBytes, getDownloadURL, deleteObject } from 'firebase/storage';
import { storage } from '../firebaseConfig';

/**
 * Upload image file to Firebase Storage
 * @param {File} file - The file to upload
 * @param {string} folder - Storage folder (e.g. 'avatars')
 * @returns {Promise<string>} Download URL
 */
export async function uploadImage(file, folder = 'avatars') {
  const fileName = `${folder}/${Date.now()}_${file.name}`;
  const storageRef = ref(storage, fileName);
  const snapshot = await uploadBytes(storageRef, file);
  // Small delay to ensure file is available (matching Flutter's whenComplete fix)
  await new Promise(resolve => setTimeout(resolve, 500));
  const url = await getDownloadURL(snapshot.ref);
  return url;
}

/**
 * Delete file from Storage by its download URL
 * @param {string} url - The download URL
 */
export async function deleteByUrl(url) {
  try {
    const storageRef = ref(storage, url);
    await deleteObject(storageRef);
  } catch (e) {
    console.warn('Storage delete failed:', e);
  }
}

File: lib/services/storage_service.dart
import 'dart:io';
import 'dart:typed_data';
import 'package:flutter/foundation.dart' show kIsWeb;
import 'package:firebase_storage/firebase_storage.dart';

/// Service upload Firebase Storage hỗ trợ cả Web (bytes) và Mobile (File).
///
/// Hiện tại chỉ dùng cho upload ảnh đại diện (avatar).
/// Audio được phát qua TTS API, không cần lưu vào Storage.
class StorageService {
  StorageService._();

  static final _storage = FirebaseStorage.instance;

  // ─── IMAGE UPLOAD ──────────────────────────────────────────

  static Future<String> uploadImageFromPath(
    String filePath, {
    String folder = 'avatars',
  }) async {
    final name =
        '${DateTime.now().millisecondsSinceEpoch}_${filePath.split(Platform.pathSeparator).last}';
    final ref = _storage.ref().child('$folder/$name');
    final snapshot = await ref.putFile(File(filePath)).whenComplete(() {});
    return await snapshot.ref.getDownloadURL();
  }

  static Future<String> uploadImageFromBytes(
    Uint8List bytes,
    String fileName, {
    String folder = 'avatars',
  }) async {
    final name = '${DateTime.now().millisecondsSinceEpoch}_$fileName';
    final ref = _storage.ref().child('$folder/$name');
    final snapshot = await ref
        .putData(bytes, SettableMetadata(contentType: 'image/jpeg'))
        .whenComplete(() {});
    return await snapshot.ref.getDownloadURL();
  }

  /// Upload ảnh chung: tự động chọn bytes (web) hoặc path (mobile).
  static Future<String> uploadImage({
    String? filePath,
    Uint8List? bytes,
    String? fileName,
    String folder = 'avatars',
  }) async {
    if (kIsWeb && bytes != null) {
      return uploadImageFromBytes(
        bytes,
        fileName ?? 'avatar.jpg',
        folder: folder,
      );
    } else if (!kIsWeb && filePath != null) {
      return uploadImageFromPath(filePath, folder: folder);
    }
    throw Exception('StorageService: Không có dữ liệu file để upload!');
  }
}
